tests/xmap_test.py

```diff
@@ -272,11 +276,204 @@ def tearDown(self):
+
+  Args:
+    sizes: dict mapping logical axis name to its corresponding size.
+
+  Returns:
+    A generator producing finitely many values, where each value is a pair in
+    which the first element is a value suitable for xmap's axis_resources
+    argument and the second element is a list of pairs with the first element
+    representing a generated physical mesh axis name and the second element
+    representing a corresponding generated mesh axis size. The generated mesh
+    names/sizes can be used to define a physical mesh in tests.
+
+  This function doesn't generate schedules which map distinct logical axis names
+  to the same parallel resource name. It only generates parallel resources; the
+  rest are implicitly left for vectorization. Parallel resource names are
+  generated by prepending an 'r' to the corresponding logical name.
+
+  Examples:
+    >>> for sched in schedules({'i': 2, 'j': 4}):
+    ...   print(sched)
+    ([], [])
+    ([('i', 'ri')], [('ri', 1)])
+    ([('i', 'ri')], [('ri', 2)])
+    ([('j', 'rj')], [('rj', 1)])
+    ([('j', 'rj')], [('rj', 2)])
+    ([('j', 'rj')], [('rj', 4)])
+    ([('j', 'rj'), ('i', 'ri')], [('rj', 1), ('ri', 1)])
+    ([('j', 'rj'), ('i', 'ri')], [('rj', 1), ('ri', 2)])
+    ([('j', 'rj'), ('i', 'ri')], [('rj', 2), ('ri', 1)])
+    ([('j', 'rj'), ('i', 'ri')], [('rj', 2), ('ri', 2)])
+    ([('j', 'rj'), ('i', 'ri')], [('rj', 4), ('ri', 1)])
+    ([('j', 'rj'), ('i', 'ri')], [('rj', 4), ('ri', 2)])
+  """
+  def divisors(n: int) -> List[int]:
+    return [m for m in range(1, n + 1) if not n % m]
+
+  for names in powerset(sizes):
+    for mesh_sizes in it.product(*(divisors(sizes[n]) for n in names)):
+      axis_resources = dict((name, 'r' + name) for name in names)
+      mesh_data = [('r' + name, size) for name, size in zip(names, mesh_sizes)]
+      yield axis_resources, mesh_data
+
+@contextmanager
+def generate_mesh(named_shape: MeshSpec) -> Generator[None, None, None]:
+  """Test utility for setting up meshes given mesh data from `schedules`."""
+  # This is similar to the `with_mesh` function above, but isn't a decorator.
+  # Could de-duplicate.
+  axis_names, shape = unzip2(named_shape)
+  size = prod(shape)
+  local_devices = list(jax.local_devices())
+  if len(local_devices) < size:
+    raise SkipTest(f"Test requires {size} local devices")
+  mesh_devices = np.array(local_devices[:size]).reshape(shape)
+  with mesh(mesh_devices, axis_names):
+    yield
+
+def schedules_from_pdot_spec(
+    spec: PdotSpec, lhs_shape: Tuple[int], rhs_shape: Tuple[int]
+    ) -> Generator[Tuple[AxisResources, MeshSpec], None, None]:
+  names = gen_axis_names(spec.num_mapped)
+  lhs_spec = dict(zip(spec.lhs_mapped, names))
+  rhs_spec = dict(zip(spec.rhs_mapped, names))
+  logical_sizes = {
+      name: shape[ax]
+      for shape, spec in [(lhs_shape, lhs_spec), (rhs_shape, rhs_spec)]
+      for ax, name in spec.items()}
```

Comment:
Ugh this part was really confusing to me, because I kept wondering what does `PdotSpec.items()` do. I've only realized that you've shadowed the name a while later 🤦  Can you please change that?

Reply:
Oops. That was an oversight. Thanks for the catch!